### PR TITLE
Add Route::headless()

### LINKED ISSUE #1: Route::headless()
### Overview

Adding the routes is easy but for even quicker testing of features.. it would be cool to: 

```
composer require headlesslaravel/headlesslaravel
```
```
php artisan headless:install
```
and just add 1 route to web.php to get started
```php
Route::headless();
```

- Automatic discover `formations` in app/http/formations and make Route::formation from them
- Automatic discover imports: if the formation ->imports() count > 0, add ->import() routes
- Automatic discover `cards` in app/http/cards and make Route::cards from them
- Automatic discover notifications, needs [route macro first](https://github.com/headlesslaravel/notifications/issues/3)

Im guessing we could just use symfony finder for this:

```php
foreach ((new Finder())->in(app_path('Http/Formations') as $file) {
    $formation = $file->someMethod();
    $resource = app($formation)->guessResourceName();
    Route::formation($formation)
        ->resource($resource);
}

foreach ((new Finder())->in(app_path('Http/Cards') as $file) {

}
```

Lets move the logic for guessing resource name to a method like above so we can call it like that

But it also allows the developer to add a method to the formation to handle guessing.

---

Maybe a config path and default to app_path if null to make it easy to test

```php
if($path = config('headless-laravel.discover-path')) {

} else {
    $path = app_path('Http/Formations')
}
```
Even though the config isnt real, it can be set during the test

Then we can call it from a test and see that it registers from tests/Fixtures

--

Lets put this in a class `Headless`

```php
Headless::formations()
Headless::cards()
```
```php
[
    UserFormation::class,
    AccountFormation::class,
]
```

### PR DESCRIPTION
**Pull request by mitulgolakiya**
closes #1

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -14,7 +14,10 @@
         "headlesslaravel/notifications": "*",
         "headlesslaravel/cards": "*",
         "headlesslaravel/metrics": "*",
-        "illuminate/console": "*"
+        "illuminate/console": "*",
+        "illuminate/routing": "*",
+        "illuminate/support": "*",
+        "composer/composer": "^2.2"
     },
```
Comment (dillingham) [inline composer.json:20]: Do we need all these dependencies?

File: src/Headless.php [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+            return $formations;
+        }
+
+        $formations = $this->getClasses($formationsPath);
+
+        foreach ($formations as $formationName) {
+            /** @var Formation $formation */
+            $formation = app($formationName);
+            Route::formation($formationName)
+                ->resource((string) $formation->guessResourceName()); // guessResourceName needs a PR to formation repo
```
Comment (dillingham) [inline src/Headless.php]:
```suggestion
                ->resource((string) $formation->guessResourceName());
```

File: src/Headless.php [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+            return $cards;
+        }
+
+        $cards = $this->getClasses($cardsPath);
+
+        foreach ($cards as $cardName) {
+            /** @var CardGroup $cardGroup */
+            $cardGroup = app($cardName);
+            /** @var Card $card */
+            Route::cards($cardGroup->guessEndpointName(), $cardName); // // guessEndpointName needs a PR to cards repo
```
Comment (dillingham) [inline src/Headless.php]:
```suggestion
            Route::cards($cardGroup->guessEndpointName(), $cardName);
```

File: src/Headless.php [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+use Composer\Autoload\ClassMapGenerator;
+use HeadlessLaravel\Cards\Card;
+use HeadlessLaravel\Cards\CardGroup;
+use HeadlessLaravel\Formations\Formation;
+use Illuminate\Support\Facades\Route;
+
+class Headless
+{
+    public function formations()
```
Comment (dillingham) [inline src/Headless.php]: Lets make `Headless::formations()` return a list of classes so its usuable for others

And lets cache it in an array of $formations,

```
if count($this->formations) !== 0

return $this->formations;

else check class map
```

Lets add that and change formations() to `routeFormations()` (which calls->formations())

File: src/Headless.php [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+
+        foreach ($formations as $formationName) {
+            /** @var Formation $formation */
+            $formation = app($formationName);
+            Route::formation($formationName)
+                ->resource((string) $formation->guessResourceName()); // guessResourceName needs a PR to formation repo
+        }
+    }
+
+    public function cards()
```
Comment (dillingham) [inline src/Headless.php]: Lets make `Headless::cards()` return a list of classes so its usuable for others

And lets cache it in an array of $cards,

```
if count($this->cards) !== 0

return $this->cards;

else check class map
```

Lets add that and change cards() to `routeCards()` (which calls->formations())

**Review by dillingham — CHANGES_REQUESTED** [review]

File: src/Headless.php [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+        }
+
+        return $classes;
+    }
+
+    public function create()
+    {
+        $this->formations();
+        $this->cards();
+    }
```
Comment (dillingham) [inline src/Headless.php:97]: if notifications is installed

```php
$this->routeNotifications()
```

File: composer.json (hunk 1 above)
Comment (mitulgolakiya) [reply composer.json:20]: We have a command in our repo, so `illuminate/console` is required.
I think we can remove `illuminate/routing`.
`illuminate/support` will be needed for different facades and service provider
`composer/composer` is needed for class mapping.

Comment (dillingham) [reply composer.json:20]: @mitulgolakiya ok thank you

File: src/Headless.php [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+        foreach ($classMap as $className => $path) {
+            $classes[] = $className;
+        }
+
+        return $classes;
+    }
+
+    public function create()
+    {
+        $this->formations();
```
Comment (dillingham) [inline src/Headless.php]:
```suggestion
        $this->routeFormations();
```

File: src/Headless.php [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+        $this->cards();
```
Comment (dillingham) [inline src/Headless.php]:
```suggestion
        $this->routeCards();
```

**Review by dillingham — CHANGES_REQUESTED** [review]

File: src/Headless.php [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+
+        $classes = [];
+        foreach ($classMap as $className => $path) {
+            $classes[] = $className;
+        }
+
+        return $classes;
+    }
+
+    public function create()
```
Comment (dillingham) [inline src/Headless.php]: lets change this to `->route()` incase headless does anything else in the future

**Pull request merged**